### DISCUSSION TITLE
Save and restore window layouts

### DISCUSSION THREAD

**Discussion by owpac** [rxhanson/Rectangle#684] [Ideas]
First of all, thank you for your work ☺️

It would be amazing to add a way to save and restore window layouts like [Moom](https://manytricks.com/moom/) or [Mosaic](https://www.lightpillar.com/mosaic.html).

What do you think about this idea?

**Comment by rxhanson** [comment]
I'm planning on adding something like this in [Hookshot](https://hookshot.app), but probably won't add it into Rectangle (in the interest of keeping Rectangle as simple as possible).

**Reply by owpac** [reply]
Didn't know about Hookshot!

However, I see 2 benefits to add this in Rectangle:
- It would be the only open source project to have this kind of feature
- This will centralize all the features in a single app without going outside the scope of Rectangle

Moreover, I think it is possible to keep this kind of feature lightweight. ☺️

**Reply by rxhanson** [reply]
Hookshot contains all of Rectangle.

I would consider a good pull request for having it in Rectangle, but would still prefer it be hidden and toggleable via terminal command and only added to the UI with popular demand. I’m pretty serious about keeping Rectangle slim. If you are interested in contributing, I can give you some entry points to look at.

**Reply by owpac** [reply]
> Hookshot contains all of Rectangle.

Oh ok, I didn't know that!

> I’m pretty serious about keeping Rectangle slim.

I can totally understand. And having this feature as an hidden and toggleable option isn't so disturbing in my opinion.

> If you are interested in contributing, I can give you some entry points to look at.

To be honest, I never did Swift at all. But I can give it a try. I'm interested to have some entry points to look at ☺️

**Reply by lucasgismondi** [reply]
I would absolutely love if this were included in Rectangle. It could be as simple as having two functions.. "Save all window positions" and "Restore all window positions" (similar to Display Maid) Maybe to add more value in Hookshot you could add different position profiles so you're not just limited to one saved layout.

I have some experience with Swift and would love to help out with this! I've got some time over the holidays.

**Reply by rxhanson** [reply]
@Owpac @lucasgismondi sounds great, I’m excited for the interest in contributing!

In my opinion, the best initial reference in the code is the TodoManager class, since that will handle all windows, as opposed to the rest of the window actions that only apply adjustments to the frontmost window. The Defaults file is Rectangle’s wrapper around  the macOS prefs storage, and the menu is built in the AppDelegate. Let me know if you have any questions. 

As an aside, feel free to implement as much or as little as you want on this feature. I don’t purposely keep Rectangle slim for the sake of beefing up Hookshot, I really just want things in Rectangle that a lot of people seriously want, because simpler is better for the vast majority of users (and another good gauge for how much people want a feature is if users are interested in implementing it themselves). My plan still would be to have it only enabled via terminal command at first though.

**Reply by lucasgismondi** [reply]
Awesome, thanks for the info. I'll try my hand at this on the weekend

Fair enough! I'll start with saving one layout for now and go from there. Will keep it behind a flag for sure 👍

**Reply by owpac** [reply]
Thanks for the info! Is there any Discord server to talk about Rectangle (and this feature)?

Not sure I can be as fast as you @lucasgismondi, but I will do my best to help ☺️

**Reply by lucasgismondi** [reply]
A discord server would be great for Rectangle, especially for collaboration. @Owpac in the meantime I wouldn't mind pairing on this and ideating over discord

**Reply by owpac** [reply]
@lucasgismondi I've sent you a message on LinkedIn ☺️

**Reply by lucasgismondi** [reply]
After some deliberation with @Owpac I'm going to step back from this feature since there are no public API's for spaces as mentioned in the Rectangle README. I would have liked windows to live in specific spaces with a specific layout. Windows being on a certain screen just won't do for my needs. Hopefully Apple can make this API public soon!
